src/Compilers/CSharp/Portable/Syntax/ArgumentSyntax.cs

```diff
@@ -1,17 +1,128 @@
+        /// </summary>
+        public ExpressionSyntax Expression
+        {
+            get
+            {
+                if (RefOrOutKeyword.Kind() != SyntaxKind.OutKeyword)
+                {
+                    return (ExpressionSyntax)ExpressionOrDeclaration;
+                }
+
+                return ExpressionOrDeclaration as ExpressionSyntax;
+            }
+        }
+
+        public ArgumentSyntax WithExpression(ExpressionSyntax expression)
+        {
+            return this.Update(this.NameColon, this.RefOrOutKeyword, expression);
+        }
+
+        /// <summary>
+        /// Get declaration of an Out Variable for the argument.
+        /// </summary>
+        public VariableDeclarationSyntax VariableDeclaration
+        {
+            get
+            {
+                if (RefOrOutKeyword.Kind() != SyntaxKind.OutKeyword)
+                {
+                    return null;
+                }
+
+                return ExpressionOrDeclaration as VariableDeclarationSyntax;
+            }
+        }
+
+        /// <summary>
+        /// Get identifier for out variable declaration, or default(SyntaxToken).
+        /// </summary>
+        internal SyntaxToken Identifier
+        {
+            get
+            {
+                if (RefOrOutKeyword.Kind() == SyntaxKind.OutKeyword && ExpressionOrDeclaration.Kind() == SyntaxKind.VariableDeclaration)
+                {
+                    return ((VariableDeclarationSyntax)ExpressionOrDeclaration).Variables.First().Identifier;
+                }
+
+                return default(SyntaxToken);
+            }
+        }
+
+        /// <summary>
+        /// Get type syntax for out variable declaration, or null.
+        /// </summary>
+        internal TypeSyntax Type
+        {
+            get
+            {
+                if (RefOrOutKeyword.Kind() == SyntaxKind.OutKeyword && ExpressionOrDeclaration.Kind() == SyntaxKind.VariableDeclaration)
+                {
+                    return ((VariableDeclarationSyntax)ExpressionOrDeclaration).Type;
+                }
+
+                return null;
+            }
+        }
+
         public ArgumentSyntax Update(NameColonSyntax nameColon, SyntaxToken refOrOutKeyword, ExpressionSyntax expression)
         {
-            return this.Update(nameColon, refOrOutKeyword, expression, type: null, identifier: default(SyntaxToken)) ;
+            return this.Update(nameColon, refOrOutKeyword, (CSharpSyntaxNode)expression) ;
         }
 
-        public ArgumentSyntax Update(NameColonSyntax nameColon, SyntaxToken refOrOutKeyword, TypeSyntax type, SyntaxToken identifier)
+        public ArgumentSyntax Update(NameColonSyntax nameColon, SyntaxToken outKeyword, VariableDeclarationSyntax declaration)
         {
-            return this.Update(nameColon, refOrOutKeyword, null, type, identifier);
+            if (outKeyword.Kind() != SyntaxKind.OutKeyword)
+            {
+                throw new ArgumentException(nameof(outKeyword));
```

Comment:
Update eventually calls factory method that validates the input and throws. Once the factory is adjusted to do the validation, this code can be removed from here. In any case, it is consistent with behavior for the rest of the syntax nodes.